File: packages/frontend/app/api/dynamo/dynamo.ts
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocument } from "@aws-sdk/lib-dynamodb";

import { DynamoGuildProps } from "@/app/types";

const stage = "prod"; // Currently UI only deploys to prod, using process.env.STAGE results in undefined being rendered

const client = new DynamoDBClient();
const documentClient = DynamoDBDocument.from(client);

const defaultProps: DynamoGuildProps = {
  acknowledgment: false,
  guild_id: "",
  region: "",
  userlist: [],
  webhook_id: "",
  webhook_url: "",
};

export const getAll = async () => {
  const res = await documentClient.scan({
    TableName: `${stage}-zoe-bot-db`,
  });

  return res.Items ?? [];
};

export const getGuild = async (guildId: string): Promise<DynamoGuildProps> => {
  const res = await documentClient.get({
    TableName: `${stage}-zoe-bot-db`,
    Key: { guild_id: BigInt(guildId) },
  });

  return res.Item
    ? { ...defaultProps, ...res.Item }
    : {
        acknowledgment: false,
        guild_id: guildId,
        region: "",
        userlist: [],
        webhook_id: "",
        webhook_url: "",
      };
};

// adduser

// deluser

// region

export const createGuild = async (
  guildId: string,
  webhookId: string,
  webhookUrl: string,
): Promise<DynamoGuildProps> => {
  try {
    await documentClient.put({
      TableName: `${stage}-zoe-bot-db`,
      Item: {
        acknowledgment: false,
        guild_id: BigInt(guildId),
        region: "NA",
        userlist: [],
        webhook_id: webhookId,
        webhook_url: webhookUrl,
      },
    });
  } catch (e) {
    return defaultProps;
  }

  return {
    acknowledgment: false,
    guild_id: guildId,
    region: "NA",
    userlist: [],
    webhook_id: webhookId,
    webhook_url: webhookUrl,
  };
};

export const destroyGuild = async (
  guild: DynamoGuildProps,
): Promise<DynamoGuildProps> => {
  try {
    await documentClient.delete({
      TableName: `${stage}-zoe-bot-db`,
      Key: { guild_id: BigInt(guild.guild_id) },
    });
  } catch (e) {
    return guild;
  }

  return defaultProps;
};

export const acknowledge = async (
  guild: DynamoGuildProps,
): Promise<DynamoGuildProps> => {
  try {
    await documentClient.put({
      TableName: `${stage}-zoe-bot-db`,
      Item: {
        acknowledgment: true,
        guild_id: BigInt(guild.guild_id),
        region: guild.region,
        userlist: guild.userlist,
        webhook_id: guild.webhook_id,
        webhook_url: guild.webhook_url,
      },
    });

    guild.acknowledgment = true;
  } catch (e) {}

  return guild;
};

File: packages/functions/src/layers/dynamo/python/dynamo.py
import boto3

class ZoeBotTable:
    def __init__(self, region, stage):
        self.region = region
        self.client = boto3.client('dynamodb', region_name=self.region)
        self.table_name = f'{stage}-zoe-bot-db'

    def get_all(self):
        return self.client.scan(TableName=self.table_name)

    def guild_exists(self, guild_id):
        response = []
        try:
            response = self.client.get_item(
                TableName=self.table_name,
                Key={'guild_id': {'N': guild_id}}
            )
        except Exception as e:
            import logging
            logging.basicConfig(level=logging.WARNING)
            logging.warning(f"guild_exists exception: {e}")
            pass
        return 'Item' in response

    def get_guild(self, guild_id):
        response = self.client.get_item(
            TableName=self.table_name,
            Key={'guild_id': {'N': guild_id}}
        )
        return response.get('Item', {})

    def create_guild(self, guild_id, webhook_id, webhook_url):
        item = {
            'guild_id': {'N': guild_id},
            'webhook_id': {'S': webhook_id},
            'webhook_url': {'S': webhook_url},
            'region': {'S': 'NA'},
            'userlist': {'L': []},
            'acknowledgment' : {'BOOL': False}
        }
        self.client.put_item(
            TableName=self.table_name,
            Item=item
        )
        return True

    def destroy_guild(self, guild_id):
        self.client.delete_item(
            TableName=self.table_name,
            Key={'guild_id': {'N': guild_id}}
        )
        return True

    def update_guild(self, guild_id, updates):  
        self.client.update_item(
            TableName=self.table_name,
            Key={'guild_id': {'N': guild_id}},
            AttributeUpdates=updates
        )
        return True

    def get_all_users(self, guild_id):
        response = self.client.get_item(
            TableName=self.table_name,
            Key={'guild_id': {'N': guild_id}}
        ).get('Item', {}).get('userlist', {}).get('L', [])
        userlist = []
        for user in response:
            puuid = list(user.get('M', {}).keys())
            userlist.extend(puuid)
        return userlist

    def user_exists(self, guild_id, puuid):
        users = self.get_all_users(guild_id)
        return puuid in users

    def add_user(self, guild_id, puuid, gameId=''):
        expression_values = {
            ':user': {'L': [{'M': {puuid: {'S': gameId}}}]}
        }
        self.client.update_item(
            TableName=self.table_name,
            Key={'guild_id': {'N': guild_id}},
            UpdateExpression='SET userlist = list_append(userlist, :user)',
            ExpressionAttributeValues=expression_values
        )
        return True

    def delete_user(self, guild_id, puuid): # TODO: rather than using index, remove by querying the puuid key
        userlist = self.get_all_users(guild_id)
        index = userlist.index(puuid)
        self.client.update_item(
            TableName=self.table_name,
            Key={'guild_id': {'N': guild_id}},
            UpdateExpression=f'REMOVE userlist[{index}]',
        )
        return True

    def update_user(self, guild_id, puuid, gameId):
        self.delete_user(guild_id, puuid)
        self.add_user(guild_id, puuid, gameId)
        return True

    def check_acknowledgment(self, guild_id):
        return self.get_guild(guild_id)['acknowledgment']['BOOL']
    
    def get_webhook(self, guild_id):
        return self.get_guild(guild_id)['webhook_id']['S']
